src/snowflake/snowpark/_internal/ast/batch.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from collections import namedtuple
+from dataclasses import dataclass
+from typing import Callable
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# TODO(SNOW-1791994): Enable pyright type checks for this file.
+
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+@dataclass
+class TrackedCallable:
+    """
+    Several Snowpark APIs that deal with stored procedures and user-defined functions accept callables as arguments.
+    This class is a pair of a callable and an ID that is used to reference it in the AST. Distinct objects get distinct IDs.
+    It is undesirable for the same callable to have multiple IDs due to constraints in other parts of the system.
+    """
+
+    var_id: int
+    func: Callable
+
+
+SerializedBatch = namedtuple("SerializedBatch", ["request_id", "batch"])
+
+
+class AstBatch:
+    """
+    A batch of AST statements. This class is used to generate AST requests.
+
+    The core statement types are:
+    - Assign: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    generate_request_id = uuid.uuid4
+
+    def __init__(self, session) -> None:
```

Comment:
Can we have all arguments and return value with type hints? Unfortunately we don't have a lint for it due to protobuf. So let's try to treat this as a best practice for now. 

Sorry I was insisted on this but the type hints actually help a lot when we deploying pandas APIs. It is especially helpful to detect bugs earlier.

Let me try to see how to enforce pyright on this later.

Reply:
+1

Reply:
Done.